### DISCUSSION TITLE
Bulk query observations and issues

### DISCUSSION THREAD

**Discussion by Rob-ModernAnalytics** [jsforce/jsforce#1446] [Q&A]
Hi,

I'm using the bulk query to download large amounts of data (I'm testing with 500,000 Salesforce records) to a CSV.  I've noticed that this uses a huge amount of memory (> 4GB) - UNLESS I comment out the on('record') handler.  If I comment out the on('record') handler then the call uses about 500MB of memory.

However, and this is a big however, if I comment out the on('record') handler then the on('end') handler is never triggered.

I don't really need to do anything with the on('record') handler as that would slow the download process, but why isn't the on('end') triggered if the on('record') doesn't exist?

Also, with the on('record') active, node never releases the memory (4GB in my test) that was used.

I had to set the --max_old_space_size=8192 to be able to download these rows.  I am using the .stream().pipe(fsStream) technique.  I thought that streaming didn't use large amounts of memory as each record was simply piped to the output file as data rows were read from Salesforce.  If I don't set the heap size, then I get an ECONNRESET error.  I know other folks have experienced similar issues.  Here is my code:

			salesForceConnection.bulk.pollTimeout = Number.MAX_VALUE;
			salesForceConnection.bulk.pollInterval = Number.MAX_VALUE;
			salesForceConnection.bulk.query(soqlStatement)
                                // Commenting the .on('record') uses very little memory but doesn't trigger the on('end')!!
				.on('record', function () {
				})
				.on('end', () => {
					console.log('end');
					return resolve(true);
				})
				.on('error', function (err) {
					console.error('error ' + err);
					return reject(err);
				})
				.stream().pipe(fsStream)
			;

**Comment by hporter** [comment]
@Rob-ModernAnalytics Any luck with this? I think I have had similar issues before, will report back if I find a solution.

Looks like there is another issue with large bulk streams as well:

https://github.com/jsforce/jsforce/issues/697

**Comment by Rob-ModernAnalytics** [comment]
Henry,

I ended up piping the records through a transform stream and putting my record and end methods in the transform stream instead of the jsforce stream.  Something like this:

			const stream = require('stream');

			// Transform stream...
			var transform = new stream.Transform();
			//transform.writable = true;
			transform.write = function(data) {
				// Send the buffer along to next pipeline item...
				this.push(data);
                                ...
                       }
			transform.end = function() {
				this.push(null);
			};
			salesForceConnection.bulk.query(soqlStatement)
				.stream().pipe(transform).pipe(fsStream)
			;
